lib/ts/recipe/session/middleware.ts

```diff
@@ -13,31 +13,54 @@
  * under the License.
  */
 import { Response, NextFunction, Request } from "express";
-import { SessionRequest } from "./types";
+import { SessionRequest, VerifySessionOptions } from "./types";
 import SessionRecipe from "./sessionRecipe";
 import { normaliseHttpMethod, sendNon200Response } from "../../utils";
 import NormalisedURLPath from "../../normalisedURLPath";
+import STError from "./error";
 
-export function verifySession(recipeInstance: SessionRecipe, antiCsrfCheck?: boolean) {
+function isRefreshAPICall(recipeInstance: SessionRecipe, request: Request): boolean {
+    let refreshTokenPath = recipeInstance.config.refreshTokenPath;
+    let method = normaliseHttpMethod(request.method);
+    let incomingPath = new NormalisedURLPath(
+        recipeInstance,
+        request.originalUrl === undefined ? request.url : request.originalUrl
+    );
+    return incomingPath.equals(refreshTokenPath) && method === "post";
+}
+
+export function verifySession(recipeInstance: SessionRecipe, options?: VerifySessionOptions | boolean) {
     // We know this should be Request but then Type
     return async (request: SessionRequest, response: Response, next: NextFunction) => {
         try {
             let method = normaliseHttpMethod(request.method);
             if (method === "options" || method === "trace") {
                 return next();
             }
-            let incomingPath = new NormalisedURLPath(
-                recipeInstance,
-                request.originalUrl === undefined ? request.url : request.originalUrl
-            );
-            let refreshTokenPath = recipeInstance.config.refreshTokenPath;
-            if (incomingPath.equals(refreshTokenPath) && method === "post") {
+            let antiCsrfCheck =
+                options !== undefined ? (typeof options === "boolean" ? options : options.antiCsrfCheck) : undefined;
+            if (isRefreshAPICall(recipeInstance, request)) {
                 request.session = await recipeInstance.refreshSession(request, response);
             } else {
                 request.session = await recipeInstance.getSession(request, response, antiCsrfCheck);
             }
             return next();
         } catch (err) {
+            /**
+             * checking:
+             *  - it should not be the refresh API
+             *  - error thrown should be either unauthorised or try refresh token
+             *  - sessionRequired parameter is set to false in options
+             */
+            if (
+                !isRefreshAPICall(recipeInstance, request) &&
+                (err.type === STError.UNAUTHORISED || err.type === STError.TRY_REFRESH_TOKEN) &&
```

Comment:
This should only be `(err.type === STError.UNAUTHORISED)` instead of `(err.type === STError.UNAUTHORISED || err.type === STError.TRY_REFRESH_TOKEN)`. Cause if it throws try refresh token, then a session does exist, and we want to refresh that.. not give undefined as `req.session`. 

There should be a test for this case. That an expired access token is used when sessionRequired is false. That should yield a try refresh token 401 error.

Reply:
Removed `TRY_REFRESH_TOKEN` part, but need to add a test